modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/DestinationFactory.java

```diff
@@ -0,0 +1,117 @@
+
+/**
+ * Factory that resolves {@link IgniteDistribution} trait, which represents logical {@link DistributionFunction} function, into its
+ * physical representation - {@link Destination} function.
+ */
+public class DestinationFactory<RowT> {
+    private final HashFunctionFactory<RowT> hashFunctionFactory;
+    private final ResolvedDependencies dependencies;
+
+    /**
+     * Constructor.
+     *
+     * @param hashFunctionFactory Hash-function factory required to resolve hash-based distributions.
+     * @param dependencies Dependencies required to resolve row value dependent distributions.
+     */
+    DestinationFactory(HashFunctionFactory<RowT> hashFunctionFactory, ResolvedDependencies dependencies) {
+        this.hashFunctionFactory = hashFunctionFactory;
+        this.dependencies = dependencies;
+    }
+
+    /**
+     * Creates a destination based on given distribution and nodes mapping.
+     *
+     * @param distribution Distribution function.
+     * @param group Target mapping.
+     * @return Destination function.
+     */
+    Destination<RowT> createDestination(IgniteDistribution distribution, ColocationGroup group) {
+        DistributionFunction function = distribution.function();
+        ImmutableIntList keys = distribution.getKeys();
+
+        switch (function.type()) {
+            case SINGLETON:
+                if (group == null || group.nodeNames() == null || group.nodeNames().size() != 1) {
+                    throw new IllegalStateException();
+                }
+
+                return new AllNodes<>(Collections.singletonList(Objects.requireNonNull(first(group.nodeNames()))));
+            case BROADCAST_DISTRIBUTED:
+                if (group == null || nullOrEmpty(group.nodeNames())) {
+                    throw new IllegalStateException();
+                }
+
+                return new AllNodes<>(group.nodeNames());
+            case RANDOM_DISTRIBUTED:
+                if (group == null || nullOrEmpty(group.nodeNames())) {
+                    throw new IllegalStateException();
+                }
+
+                return new RandomNode<>(group.nodeNames());
+            case HASH_DISTRIBUTED: {
+                if (group == null || nullOrEmpty(group.assignments()) || keys.isEmpty()) {
+                    throw new IllegalStateException();
+                }
+
+                List<List<String>> assignments = Commons.transform(group.assignments(), v -> Commons.transform(v, NodeWithTerm::name));
+
+                if (IgniteUtils.assertionsEnabled()) {
+                    for (List<String> assignment : assignments) {
+                        assert nullOrEmpty(assignment) || assignment.size() == 1;
+                    }
+                }
+
+                if (function.affinity()) {
+                    int tableId = ((AffinityDistribution) function).tableId();
+
+                    TableDescriptor tableDescriptor = dependencies.updatableTable(tableId).descriptor();
```

Comment:
I think there is a ticket to remove updatable table from `ResolvedDependencies`, if there is, it would be better to hide `TableDescriptor` behind a method that returns `TableDescriptor ` by `tableId`. What do you think?